### PR TITLE
feat(Icon): rewrite to TypeScript and hooks

### PR DESCRIPTION
**Pull request by tujoworker**
Will solve the issue described in this PR #3699

Even all the tests and types are running green, it needs extensive manual testing.

### CODE REVIEW

**Comment by langz** [comment]
Wow, that was fast 🤯 🚀

**Review by joakbjerk — APPROVED** [review]
Looks good ⭐️

File: packages/dnb-eufemia/src/components/icon/Icon.tsx [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+  } = prepareIcon(props, context)
+  const icon = iconProp ?? children
+
+  if (!icon) {
+    return null
+  }
+
+  const IconContainer = prerenderIcon({ icon, size, alt })
+
+  // make sure we return an empty span if we don't could get the icon
```
Comment (langz) [inline packages/dnb-eufemia/src/components/icon/Icon.tsx:146]:
```suggestion
  // make sure we return an empty span if we couldn't get the icon
```

File: packages/dnb-eufemia/src/components/icon/Icon.tsx [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    IconProps,
+    'icon'
+  > & { icon: IconType }
+
+  let sizeAsInt: ValidIconNumericSize | -1 = null
+  let sizeAsString = null
+
+  // if there is no size, check if we can find the actual size in the name
+  if (!size || size === DefaultIconSize) {
+    // get the icon name - we use is for several things
```
Comment (langz) [inline packages/dnb-eufemia/src/components/icon/Icon.tsx:175]:
```suggestion
    // get the icon name - we use it for several things
```

File: packages/dnb-eufemia/src/components/icon/Icon.tsx [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+      }
+    }
+  }
+
+  // if size is defined as a string, find the size number
+  else if (typeof size === 'string' && !(parseFloat(size) > 0)) {
+    sizeAsInt =
+      ListDefaultIconSizes.filter(([key]) => key === size)?.[0]?.[1] ?? -1
+
+    // of if the size is a default size defined as a string
```
Comment (langz) [inline packages/dnb-eufemia/src/components/icon/Icon.tsx:217]:
```suggestion
    // or if the size is a default size defined as a string
```

File: packages/dnb-eufemia/src/components/icon/Icon.tsx [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+    if (ValidIconType.includes(size as ValidIconType)) {
+      sizeAsString = size
+    }
+  }
+
+  // check if the size is given as a number, and if is a default size
```
Comment (langz) [inline packages/dnb-eufemia/src/components/icon/Icon.tsx:223]:
```suggestion
  // check if the size is given as a number, and if it's a default size
```

**Review by langz — APPROVED** [review]
Quickly gave it a read-through, will try to do some more manually testing during the day, and report back 🙇

**Pull request merged**